### PR TITLE
Cknieling/array attribute UI

### PR DESCRIPTION
**Pull request by chr11115**
- Add UI handles to edit array attributes

### CODE REVIEW

File: src/palladio/SOPAssign.cpp [hunk 1]
```diff
@@ -154,6 +152,72 @@ void updateUIDefaultValues(SOPAssign* node, const std::wstring& style,
...
+									if (i >= wStringValues.size() || !parmInst->isDefault())
+										continue;
+
+									const UT_StringHolder stringValue(toOSNarrowFromUTF16(wStringValues[i]));
+
+									node->setString(stringValue, CH_StringMeaning::CH_STRING_LITERAL,
+									                parmInstToken.c_str(), 0, time);
+									break;
+								}
+							}
```
Comment (mistafunk) [inline src/palladio/SOPAssign.cpp:214]: You are doing three things at once here... getting the default array size, converting the values into HDK types and setting the individual array values in the parm... especially interleaving `it` and `i` is hard to read...

try to separate this into functions which do one thing at a time

File: src/palladio/SOPAssign.cpp [hunk 2]
```diff
@@ -462,6 +465,27 @@ bool evaluateDefaultRuleAttributes(SOPAssign* node, const GU_Detail* detail, Sha
 	return true;
 }
 
+template<typename T>
+std::pair<UT_ValArray<T>, bool> getArrayFromMultiParm(const SOPAssign* const node, const PRM_Parm& parm, PRM_Type expectedType, std::function<T(const SOPAssign*, const PRM_Parm*)> eval) {
```
Comment (mistafunk) [inline src/palladio/SOPAssign.cpp:469]: consider replacing std::function with a template argument (std::function has runtime overhead and is more verbose)

```
template<typename T, typename E>
std::pair<UT_ValArray<T>, bool> getArrayFromMultiParm(const SOPAssign* const node, const PRM_Parm& parm, PRM_Type expectedType, E eval) {
```

File: src/palladio/NodeSpareParameter.cpp [hunk 3]
```diff
@@ -159,6 +159,63 @@ void addStringParm(OP_Node* node, const std::wstring& id, const std::wstring& na
...
+
+	PRM_Template itemTemplate[] = {PRM_Template(PRM_STRING, 1, &itemName, PRMzeroDefaults), PRM_Template()};
+
+	PRM_Template myTemplateList[] = {PRM_Template(PRM_MULTITYPE_LIST, itemTemplate, 0, &templateName, &defaultVal,
+	                                              nullptr, &PRM_SpareData::multiStartOffsetZero),
+	                                 PRM_Template()};
+
+	addParmsFromTemplateArray(node, myTemplateList, parentFolders);
+}
+
```
Comment (mistafunk) [inline src/palladio/NodeSpareParameter.cpp:221]: these three new functions look very similar, could you unify them with a template?

**Review by mistafunk — CHANGES_REQUESTED** [review]

File: src/palladio/SOPAssign.cpp [hunk 4]
```diff
@@ -440,7 +440,8 @@ void SOPAssign::updateDefaultCGAAttributes(const ShapeData& shapeData) {
 				case prt::AttributeMap::PT_FLOAT_ARRAY: {
 					size_t arr_length = 0;
 					const double* doubleArray = defaultRuleAttributes->getFloatArray(key, &arr_length);
-					std::vector<double> doubleVec(arr_length);
+					std::vector<double> doubleVec;
+					doubleVec.reserve(arr_length);
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/SOPAssign.cpp:532]: done

File: src/palladio/SOPAssign.cpp [hunk 5]
```diff
@@ -709,7 +709,38 @@ void SOPAssign::buildUI(GU_Detail* detail, ShapeData& shapeData, const ShapeConv
 						break;
 					}
 				}
-
+				break;
+			}
+			case prt::AnnotationArgumentType::AAT_BOOL_ARRAY: {
+				const bool isDefaultValBoolArray = (defaultValIt->second.index() == 5);
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/SOPAssign.cpp]: using std::holds_alternative<type> instead of index checks instead

File: src/palladio/SOPAssign.cpp (hunk 1 above)
Comment (chr11115) [reply src/palladio/SOPAssign.cpp]: done

File: src/palladio/SOPAssign.cpp (hunk 2 above)
Comment (chr11115) [reply src/palladio/SOPAssign.cpp]: done

File: src/palladio/SOPAssign.cpp [hunk 6]
```diff
@@ -398,6 +465,28 @@ bool evaluateDefaultRuleAttributes(SOPAssign* node, const GU_Detail* detail, Sha
...
+	UT_ValArray<T> valArray(parmCount);
+	for (int i = 0; i < parmCount; ++i) {
+		PRM_Parm* parmInst = parm.getMultiParm(i);
+		if (parmInst == nullptr || parmInst->getType().getBasicType() != expectedType)
+			break;
+		if (!parmInst->isDefault())
+			skipAttr = false;
+
+		T value;
+		if (parmInst)
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/SOPAssign.cpp:483]: removed.

File: src/palladio/SOPAssign.cpp [hunk 7]
```diff
@@ -742,10 +947,16 @@ void SOPAssign::opChanged(OP_EventType reason, void* data) {
...
 	if (reason == OP_PARM_CHANGED) {
 		// parm index is directly stored in the void*, casting to size_t to avoid compiler warnings
-		const PRM_Parm& parm = getParm((size_t)data);
+		const PRM_ParmList* parmList = getParmList();
+		if (parmList == nullptr)
+			return;
 
-		if (parm.isSpareParm()) {
-			const PRM_Type currParmType = parm.getType();
+		const PRM_Parm* const parmPtr = parmList->getParmPtr((size_t)data);
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/SOPAssign.cpp:954]: changed accordingly

File: src/palladio/SOPAssign.cpp [hunk 8]
```diff
@@ -588,27 +617,75 @@ void SOPAssign::updatePrimitiveAttributes(GU_Detail* detail) {
 					break;
 				}
 				case PRM_Type::PRM_BasicType::PRM_BASIC_FLOAT: {
-					switch (currParmType.getFloatType()) {
-						case PRM_Type::PRM_FLOAT_NONE: {
-							const double floatValue = evalFloat(&parm, 0, time);
+					if (parm.getMultiType() == PRM_MultiType::PRM_MULTITYPE_LIST) {
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/SOPAssign.cpp:621]: Separated into distinct and obvious helper functions

File: src/palladio/SOPAssign.cpp [hunk 9]
```diff
@@ -612,11 +635,7 @@ void SOPAssign::updatePrimitiveAttributes(GU_Detail* detail) {
 				case PRM_Type::PRM_BasicType::PRM_BASIC_FLOAT: {
 					if (parm.getMultiType() == PRM_MultiType::PRM_MULTITYPE_LIST) {
 						if (std::holds_alternative<std::vector<bool>>(it->second)) {
-							const std::optional<UT_Int32Array>& boolArray = getArrayFromMultiParm<int32>(
-							        this, parm, PRM_Type::PRM_BASIC_ORDINAL,
-							        [time](const SOPAssign* node, const PRM_Parm* parmInst) {
-								        return node->evalInt(parmInst, 0, time);
-							        });
+							const std::optional<UT_Int32Array>& boolArray = getBoolArrayFromParm(this, parm, time);
```
Comment (mistafunk) [inline src/palladio/SOPAssign.cpp:862]: this looks dangerous, you take a reference on a temporary return value. Better use return-by-value, the compiler should be able to "move" the array (return value optimziation, RVO).

also two more cases below

File: src/palladio/Utils.cpp [hunk 10]
```diff
@@ -69,6 +69,13 @@ std::basic_string<C> callAPI(FUNC f, size_t initialSize) {
 
 } // namespace
 
+std::vector<const wchar_t*> toPtrVec(const std::vector<std::wstring>& wsv) {
+	std::vector<const wchar_t*> pw(wsv.size());
+	for (size_t i = 0; i < wsv.size(); i++)
```
Comment (mistafunk) [inline src/palladio/Utils.cpp:87]: classic use case for std::transform, also see https://github.com/Esri/serlio/blob/ab8f5e675dd164285af62f28603664e579a5e1fa/src/serlio/utils/Utilities.h#L87

**Review by mistafunk — CHANGES_REQUESTED** [review]
Almost good. See inline comments. Afterwards feel free to directly merge.

File: src/palladio/SOPAssign.cpp (hunk 9 above)
Comment (chr11115) [reply src/palladio/SOPAssign.cpp]: fixed

**Pull request merged**